python/feathub/processors/spark/spark_table.py

```diff
@@ -0,0 +1,120 @@
+from feathub.feature_tables.sinks.print_sink import PrintSink
+from feathub.processors.spark.spark_job import SparkJob
+from feathub.processors.spark.spark_types_utils import (
+    to_feathub_schema,
+    to_feathub_type,
+)
+from feathub.table.schema import Schema
+from feathub.table.table import Table
+from feathub.table.table_descriptor import TableDescriptor
+
+if typing.TYPE_CHECKING:
+    from feathub.processors.spark.spark_processor import SparkProcessor
+
+
+class SparkTable(Table):
+    """
+    The implementation of Feathub Table for Spark.
+
+    It can only be instantiated and processed by SparkProcessor. It converts the feature
+    to native Spark DataFrame internally with the SparkDataFrameBuilder.
+    """
+
+    def __init__(
+        self,
+        spark_processor: "SparkProcessor",
+        feature: TableDescriptor,
+    ) -> None:
+        """
+        Instantiate the SparkTable.
+
+        :param spark_processor: The SparkProcessor that instantiate this SparkTable.
+        :param feature: Describes the features to be included in the table. If it is a
+                        FeatureView, it must be resolved.
+        """
+        super().__init__(feature.timestamp_field, feature.timestamp_format)
+        self.spark_processor = spark_processor
+        self.feature = feature
+
+    def get_schema(self) -> Schema:
+        return to_feathub_schema(self._get_spark_dataframe(self.feature).schema)
+
+    def to_pandas(self, force_bounded: bool = False) -> pd.DataFrame:
+        feature = self.feature
+        if not feature.is_bounded():
+            if not force_bounded:
+                raise FeathubException(
+                    "Unbounded table cannot be converted to Pandas DataFrame. You can "
+                    "set force_bounded to True to convert the Table to DataFrame."
+                )
+            feature = feature.get_bounded_view()
+
+        dataframe = self._get_spark_dataframe(feature)
+
+        field_names = []
+        field_types = []
+        for field in dataframe.schema.fields:
+            field_names.append(field.name)
+            field_types.append(to_numpy_dtype(to_feathub_type(field.dataType)))
+
+        results = dataframe.collect()
+
+        data: typing.Dict[str, typing.List[typing.Any]] = defaultdict(list)
+        for row in results:
+            for name, value in zip(field_names, row):
+                data[name].append(value)
+
+        values = [value for _, value in data.items()]
+        return pd.DataFrame(
+            {
+                field_names[i]: pd.Series(values[i], dtype=field_types[i])
+                for i in range(len(field_names))
+            }
+        )
+
+    def execute_insert(
+        self,
+        sink: FeatureTable,
+        ttl: Optional[timedelta] = None,
+        allow_overwrite: bool = False,
+    ) -> SparkJob:
+        dataframe = self._get_spark_dataframe(self.feature)
+
+        if isinstance(sink, FileSystemSink):
```

Comment:
Would it be simpler to put logic of handling various FeatureView/Source/Sink in the `SparkDataFrameBuilder`?

Reply:
I have added `spark/dataframe_builder/source_sink_utils.py` to achieve this goal. Please check if this is good enough.

Reply:
It seems better to follow the approach used by `FlinkTable#execute_insert`. This method invokes `FlinkProcessor#materialize_features`, which further invokes `FlinkJobSubmitter#submit`.

In the case here, I think `SparkTable#execute_insert` can invoke `SparkProcessor#materialize_features`, which can call SparkDataFrameBuilder APIs as appropriate.

Reply:
The current codes have exposed `FlinkProcessor`'s implementation details in `FlinkTable` (like `self.flink_processor.flink_table_builder`), so I'm not sure `FlinkTable#execute_insert` has a good enough structure. Besides, `SparkTable#get_schema` and `SparkTable#to_pandas` still need to use `dataframe_builder`, instead of some public API of `FlinkProcessor`.